cmd/nerdctl/cp.go

```diff
@@ -0,0 +1,646 @@
+type fileSpec struct {
+	Container string
+	Path      string
+}
+
+func copyToContainer(ctx context.Context, cmd *cobra.Command, copyConfig cpConfig) (err error) {
+
+	// Init tar Pipe is the first step to process copy to container
+	t := initTarPipe(false, &copyConfig)
+
+	if _, err := os.Stat(t.src); err != nil {
+		return fmt.Errorf("%s doesn't exist in local filesystem", t.src)
+	}
+
+	// TarPipe destination should exist in the container and it should be a directory
+	destination, err := securejoin.SecureJoin(filepath.Join("/proc/", fmt.Sprint(t.option.containerPid), "/root"), t.dst)
+	if err != nil {
+		return err
+	}
+	testDCmd := []string{"-d", destination}
+
+	if err := exec.Command("test", testDCmd...).Run(); err != nil {
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			if c := exiterr.ExitCode(); c == 1 {
+				return fmt.Errorf("%s should exists and should be a directory", t.dst)
+			}
+		}
+	}
+
+	go func(*TarPipe) {
+		defer t.outStream.Close()
+		makeTar(cmd, t)
+	}(t)
+
+	// check if tar binary exists in the container
+	tarBinary, err := copyutil.TarBinary()
+	if err != nil {
+		hint := fmt.Sprintf("`tar` binary needs to be installed")
+		return fmt.Errorf(hint+": %w", err)
+	}
+
+	noPreserve, err := cmd.Flags().GetBool("no-preserve")
+	if err != nil {
+		return err
+	}
+
+	compress, err := cmd.Flags().GetBool("compress")
+	if err != nil {
+		return err
+	}
+
+	gzipFlag := ""
+
+	if compress {
+		gzipFlag = "z"
+	}
+
+	var tarArgs []string
+	if noPreserve {
+		tarArgs = []string{"--no-same-permissions", "--no-same-owner", fmt.Sprintf("-x%smf", gzipFlag), "-"}
+	} else {
+		tarArgs = []string{fmt.Sprintf("-x%smf", gzipFlag), "-"}
+	}
+
+	dstDir, _ := copyutil.SplitPathDirEntry(filepath.Clean(t.dst))
+	var target string
+	if len(dstDir) > 0 {
+		target, err = securejoin.SecureJoin(filepath.Join("/proc/", fmt.Sprint(t.option.containerPid), "/root"), dstDir)
+		if err != nil {
+			return err
+		}
+	} else {
+		target, err = securejoin.SecureJoin(filepath.Join("/proc/", fmt.Sprint(t.option.containerPid), "/root"), "")
+		if err != nil {
+			return err
+		}
+	}
+
+	tarArgs = append(tarArgs, "-C", target)
+	command := exec.Command(tarBinary, tarArgs...)
+	stdin, err := command.StdinPipe()
+	if err != nil {
+		return err
+	}
+
+	go func(*TarPipe) error {
+
+		defer stdin.Close()
+		isquite, err := cmd.Flags().GetBool("quite")
+		if err != nil {
+			return err
+		}
+
+		var writer io.Writer
+		writer = stdin
+
+		if !isquite {
+			bar := progressbar.DefaultBytes(
+				-1,
+				"uploading",
+			)
+
+			writer = io.MultiWriter(stdin, bar)
+		}
+
+		if _, err := io.Copy(writer, t.reader); err != nil {
+			return err
+		}
+
+		return nil
+	}(t)
+
+	out, err := command.CombinedOutput()
+	if err != nil {
+		logrus.Errorf("failed to exec: %v: %v", string(out), err)
+	}
+
+	return nil
+}
+
+func makeTar(cmd *cobra.Command, t *TarPipe) error {
+
+	// making Tar flow:
+	// tar --> gzip --> buf
+
+	compress, err := cmd.Flags().GetBool("compress")
```

Comment:
Why do you need to compress?
It is probably slower

Reply:
removed